File: Reference Bots/Java/ReferenceBot/src/main/java/za/co/entelect/challenge/domain/state/PlayerMap.java
package za.co.entelect.challenge.domain.state;

import za.co.entelect.challenge.domain.command.Point;
import za.co.entelect.challenge.domain.command.direction.Direction;

import java.util.ArrayList;

public class PlayerMap {

    public BattleshipPlayer Owner;

    public ArrayList<Cell> Cells;

    public int MapWidth;

    public int MapHeight;

    public Cell getCellAt(int x, int y) {
        return Cells.stream().filter(p -> p.X == x && p.Y == y).findFirst().get();
    }

    public Cell getAdjacentCell(Cell cell, Direction direction) {
        if (cell == null) {
            return null;
        }

        switch (direction) {

            case North:
                return getCellAt(cell.X, cell.Y + 1);
            case East:
                return getCellAt(cell.X + 1, cell.Y);
            case South:
                return getCellAt(cell.X, cell.Y - 1);
            case West:
                return getCellAt(cell.X - 1, cell.Y);
            default:
                throw new IllegalArgumentException(String.format("The direction passed %s does not exist", direction));
        }
    }

    public ArrayList<Cell> getAllCellsInDirection(Point startLocation, Direction direction, int length) {
        Cell startCell = getCellAt(startLocation.getX(), startLocation.getY());
        ArrayList<Cell> cells = new ArrayList<>();
        cells.add(startCell);

        if (startCell == null) {
            return cells;
        }

        for (int i = 1; i < length; i++) {
            Cell nextCell = getAdjacentCell(startCell, direction);
            if (nextCell == null) {
                throw new IllegalArgumentException("Not enough cells for the requested length");
            }

            cells.add(nextCell);
            startCell = nextCell;
        }

        return cells;
    }

    public boolean hasCellsForDirection(Point startLocation, Direction direction, int length) {
        Cell startCell = getCellAt(startLocation.getX(), startLocation.getY());

        if (startCell == null) {
            return false;
        }

        for (int i = 1; i < length; i++) {
            Cell nextCell = getAdjacentCell(startCell, direction);
            if (nextCell == null) {
                return false;
            }
            startCell = nextCell;
        }
        return true;
    }
}

File: Reference Bots/C#/ReferenceBot/Domain/State/PlayerMap.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using ReferenceBot.Domain.Command;

namespace ReferenceBot.Domain.State
{
    public class PlayerMap
    {
        [JsonProperty]
        public BattleshipPlayer Owner { get; set; }
        [JsonProperty]
        public List<Cell> Cells { get; set; }
        [JsonProperty]
        public int MapWidth { get; set; }
        [JsonProperty]
        public int MapHeight { get; set; }

        public Cell GetCellAt(int x, int y)
        {
            return Cells.FirstOrDefault(p => p.X == x && p.Y == y);
        }

        public Cell GetAdjacentCell(Cell cell, Direction direction)
        {
            if (cell == null)
                return null;

            switch (direction)
            {
                case Direction.North:
                    return GetCellAt(cell.X, cell.Y + 1);
                case Direction.South:
                    return GetCellAt(cell.X, cell.Y - 1);
                case Direction.West:
                    return GetCellAt(cell.X - 1, cell.Y);
                case Direction.East:
                    return GetCellAt(cell.X + 1, cell.Y);
                default:
                    throw new ArgumentOutOfRangeException(nameof(direction), direction, null);
            }
        }

        public List<Cell> GetAllCellsInDirection(Point startLocation, Direction direction, int length)
        {
            var startCell = GetCellAt(startLocation.X, startLocation.Y);
            var cells = new List<Cell>() {startCell};

            if (startCell == null)
                return cells;

            for (int i = 1; i < length; i++)
            {
                var nextCell = GetAdjacentCell(startCell, direction);
                if(nextCell == null)
                    throw new ArgumentException("Not enough cells for requested length");

                cells.Add(nextCell);
                startCell = nextCell;
            }
            
            return cells;
        }

        public bool HasCellsForDirection(Point startLocation, Direction direction, int length)
        {
            var startCell = GetCellAt(startLocation.X, startLocation.Y);

            if (startCell == null)
                return false;

            for (int i = 1; i < length; i++)
            {
                var nextCell = GetAdjacentCell(startCell, direction);
                if (nextCell == null)
                    return false;
                
                startCell = nextCell;
            }

            return true;
        }
    }
}
